core/src/ast/attrs.rs

```diff
@@ -1,10 +1,201 @@
 
-pub(crate) fn extract_cfg_attrs(attrs: &[Attribute]) -> impl Iterator<Item = String> + '_ {
-    attrs
-        .iter()
-        .filter(|&a| a.path().is_ident("cfg"))
-        .map(|a| quote::quote!(#a).to_string())
+/// The list of attributes on a type
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Default)]
+pub struct Attrs {
+    pub cfg: Vec<Attribute>,
+    pub attrs: Vec<DiplomatAttr>,
+}
+
+impl Attrs {
+    fn add_attr(&mut self, attr: Attr) {
+        match attr {
+            Attr::Cfg(attr) => self.cfg.push(attr),
+            Attr::DiplomatAttr(attr) => self.attrs.push(attr),
+        }
+    }
+
+    /// Merge attributes that should be inherited from the parent
+    pub(crate) fn merge_parent_attrs(&mut self, other: &Attrs) {
+        self.cfg.extend(other.cfg.iter().cloned())
+    }
+    pub(crate) fn add_attrs(&mut self, attrs: &[Attribute]) {
+        for attr in syn_attr_to_ast_attr(attrs) {
+            self.add_attr(attr)
+        }
+    }
+    pub(crate) fn from_attrs(attrs: &[Attribute]) -> Self {
+        let mut this = Self::default();
+        this.add_attrs(attrs);
+        this
+    }
+}
+
+impl From<&[Attribute]> for Attrs {
+    fn from(other: &[Attribute]) -> Self {
+        Self::from_attrs(other)
+    }
+}
+
+enum Attr {
+    Cfg(Attribute),
+    DiplomatAttr(DiplomatAttr),
+    // More goes here
+}
+
+fn syn_attr_to_ast_attr(attrs: &[Attribute]) -> impl Iterator<Item = Attr> + '_ {
+    let cfg_path: syn::Path = syn::parse_str("cfg").unwrap();
+    let dattr_path: syn::Path = syn::parse_str("diplomat::attr").unwrap();
+    attrs.iter().filter_map(move |a| {
+        if a.path() == &cfg_path {
+            Some(Attr::Cfg(a.clone()))
+        } else if a.path() == &dattr_path {
+            Some(Attr::DiplomatAttr(
+                a.parse_args()
+                    .expect("Failed to parse malformed diplomat::attr"),
+            ))
+        } else {
+            None
+        }
+    })
+}
+
+impl Serialize for Attrs {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // 3 is the number of fields in the struct.
+        let mut state = serializer.serialize_struct("Attrs", 1)?;
+        let cfg: Vec<_> = self
+            .cfg
+            .iter()
+            .map(|a| quote::quote!(#a).to_string())
+            .collect();
+        state.serialize_field("cfg", &cfg)?;
+        state.end()
+    }
+}
+
+/// A `#[diplomat::attr(...)]` attribute
+///
+/// Its contents must start with single element that is a CFG-expression
+/// (so it may contain `foo = bar`, `foo = "bar"`, `ident`, `*` atoms,
+/// and `all()`, `not()`, and `any()` combiners), and then be followed by one
+/// or more backend-specific attributes, which can be any valid meta-item
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize)]
+pub struct DiplomatAttr {
+    pub cfg: DiplomatAttrCfg,
+    #[serde(serialize_with = "serialize_meta")]
+    pub attr: Meta,
```

Comment:
Thought/Suggestion: name this `meta` to be the same as the field in `syn::Attribute` which is the other variant of `Attr`?